polygon/blocksync/block_downloader.go

```diff
@@ -0,0 +1,132 @@
+package blocksync
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ledgerwatch/erigon/core/types"
+)
+
+func NewBlockDownloader(p2pLayer p2pLayer, dbLayer dbLayer, heimdallLayer heimdallLayer) *BlockDownloader {
+	return &BlockDownloader{
+		p2pLayer:      p2pLayer,
+		dbLayer:       dbLayer,
+		heimdallLayer: heimdallLayer,
+	}
+}
+
+type BlockDownloader struct {
+	p2pLayer      p2pLayer
+	dbLayer       dbLayer
+	heimdallLayer heimdallLayer
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingCheckpoints(ctx, fromBlockNum, false)
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingMilestones(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingMilestones(ctx, fromBlockNum, false)
+}
+
+func (bd BlockDownloader) BackwardDownloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingCheckpoints(ctx, fromBlockNum, true)
+}
+
+func (bd BlockDownloader) BackwardDownloadUsingMilestones(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingMilestones(ctx, fromBlockNum, true)
+}
+
+func (bd BlockDownloader) downloadUsingMilestones(ctx context.Context, fromBlockNum uint64, reverse bool) error {
+	milestones, err := bd.heimdallLayer.FetchMilestones(fromBlockNum)
+	if err != nil {
+		return err
+	}
+
+	err = bd.downloadUsingStatePoints(ctx, statePointListFromMilestones(milestones), reverse)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bd BlockDownloader) downloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64, reverse bool) error {
+	checkpoints, err := bd.heimdallLayer.FetchCheckpoints(fromBlockNum)
+	if err != nil {
+		return err
+	}
+
+	err = bd.downloadUsingStatePoints(ctx, statePointListFromCheckpoints(checkpoints), reverse)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bd BlockDownloader) downloadUsingStatePoints(ctx context.Context, spl statePointList, reverse bool) error {
+	var statePointsBatch statePointList
+	for len(spl) > 0 {
+		peerCount := bd.p2pLayer.PeerCount()
+		statePointsBatch, spl = spl.extractBatch(peerCount, reverse)
+
+		blocksBatchChannels := make([]chan []*types.Block, len(statePointsBatch))
+		errChannels := make([]chan error, len(statePointsBatch))
+		maxStatePointLength := 0
+		for i, sp := range statePointsBatch {
+			// TODO generic max func
+			length := sp.length()
+			if length > maxStatePointLength {
+				maxStatePointLength = length
+			}
+
+			blocksBatchChannels[i], errChannels[i] = bd.downloadBlockRange(sp.startBlock, sp.endBlock, i%peerCount)
+		}
+
+		blocks := make([]*types.Block, 0, maxStatePointLength*peerCount)
+		for i := 0; i < len(blocksBatchChannels); i++ {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case blocksBatch := <-blocksBatchChannels[i]:
+				blocks = append(blocks, blocksBatch...)
```

Comment:
In my mind WriteBlocks() assumes that the blocks in its argument are sorted. With this assumption we either need to keep `blocks` sorted, or call WriteBlocks() right here with the batch.
Let's discuss the DB and decide if we're ok to allow random insertions.

Reply:
they are in sorted order btw, we iterate over blocksBatchChannels to get the batches of blocks in sequential order so batches are sequential and the blocks inside these batches are sequential too